File: api/typescript/src/core/lib/enclaves/kurtosis_yaml.ts
import * as fs from "fs"
import {err, ok, Result} from "neverthrow";
import * as yaml from "js-yaml";
import {KURTOSIS_YAML_FILENAME} from "./enclave_context";

const PACKAGES_URL = "https://docs.kurtosis.com/concepts-reference/packages";

export class KurtosisYaml {
    constructor(
        public readonly  name: string,
    ){}
}

const UTF8_ENCODING = "utf-8";

export async  function parseKurtosisYaml(kurtosisYamlFilepath: string): Promise<Result<KurtosisYaml, Error>> {
    // check if the yml file actually exists
    if (!fs.existsSync(kurtosisYamlFilepath)) {
        return err(new Error(`Couldn't find a '${KURTOSIS_YAML_FILENAME}' in the root of the package at '${kurtosisYamlFilepath}'. Packages are expected to have a '${KURTOSIS_YAML_FILENAME}' at root; have a look at '${PACKAGES_URL}' for more`))
    }

    let kurtosisYamlFile: string
    try {
        kurtosisYamlFile = fs.readFileSync(kurtosisYamlFilepath, UTF8_ENCODING)
    } catch(error) {
        return err(new Error(
            `An error occurred while reading the '${KURTOSIS_YAML_FILENAME}' file at '${kurtosisYamlFilepath}'`
        ));
    }

    let parsedYAML: KurtosisYaml
    try {
        parsedYAML = (yaml.load(kurtosisYamlFile) as KurtosisYaml)
    } catch(error) {
        return err(new Error(
            `"An error occurred while parsing the '${KURTOSIS_YAML_FILENAME}' file at '${kurtosisYamlFilepath}'`
        ));
    }

    if ( parsedYAML.name === null || parsedYAML.name === "") {
        return err(new Error(`Field 'name', which is the Starlark package's name, in '${KURTOSIS_YAML_FILENAME}' needs to be set and cannot be empty`))
    }

    return ok(parsedYAML)
}

File: api/golang/core/lib/enclaves/kurtosis_yaml.go
package enclaves

import (
	"github.com/go-yaml/yaml"
	"github.com/kurtosis-tech/stacktrace"
	"os"
)

const (
	packagesUrl = "https://docs.kurtosis.com/concepts-reference/packages"
)

// fields are public because it's needed for YAML decoding
type KurtosisYaml struct {
	PackageName           string            `yaml:"name"`
	PackageDescription    string            `yaml:"description"`
	PackageReplaceOptions map[string]string `yaml:"replace"`
}

func NewKurtosisYaml(packageName string, packageDescription string, packageReplaceOptions map[string]string) *KurtosisYaml {
	return &KurtosisYaml{PackageName: packageName, PackageDescription: packageDescription, PackageReplaceOptions: packageReplaceOptions}
}

func ParseKurtosisYaml(kurtosisYamlFilepath string) (*KurtosisYaml, error) {
	kurtosisYamlContents, err := os.ReadFile(kurtosisYamlFilepath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, stacktrace.NewError("Couldn't find a '%v' in the root of the package at '%v'. Packages are expected to have a '%v' at root; have a look at '%v' for more", kurtosisYamlFilename, kurtosisYamlFilepath, kurtosisYamlFilename, packagesUrl)
		}
		return nil, stacktrace.Propagate(err, "An error occurred while reading the '%v' file at '%v'", kurtosisYamlFilename, kurtosisYamlFilepath)
	}

	var kurtosisYaml KurtosisYaml
	err = yaml.UnmarshalStrict(kurtosisYamlContents, &kurtosisYaml)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred while parsing the '%v' file at '%v'", kurtosisYamlFilename, kurtosisYamlFilepath)
	}

	if kurtosisYaml.PackageName == "" {
		return nil, stacktrace.NewError("Field 'name', which is the Starlark package's name, in %v needs to be set and cannot be empty", kurtosisYamlFilename)
	}

	return &kurtosisYaml, nil
}
